### PR TITLE
SACD Discrete Soft Actor Critic

### PR DESCRIPTION
**Pull request by splatter96**
This PR introduces the Soft Actor Critic for discrete actions (SACD) algorithm.

## Description
This PR implements the SAC-Discrete algorithm as described in this paper https://arxiv.org/abs/1910.07207. This implementation borrows code from the papers original implementation (https://github.com/p-christ/Deep-Reinforcement-Learning-Algorithms-with-PyTorch) as well as provided by the issues author who requested this feature in stable baselines (https://github.com/toshikwa/sac-discrete.pytorch)

## Context

- [x] I have raised an issue to propose this change ([required](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/master/CONTRIBUTING.md))
- [x] Original issue in the stable baselines repo https://github.com/DLR-RM/stable-baselines3/issues/157

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] The functionality/performance matches that of the source (**required** for new training algorithms or training-related features).
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have included an example of using the feature (*required for new features*).
- [ ] I have included baseline results (**required** for new training algorithms or training-related features).
- [x] I have updated the documentation accordingly.
- [ ] I have updated the changelog accordingly (**required**).
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

**Comment by araffin** [comment]
Hello,
thanks for the PR =)

>  The functionality/performance matches that of the source (required for new training algorithms or training-related features).

please don't forget that part (see contributing guide).
I think there are discussion about the results here too: https://github.com/vwxyzjn/cleanrl/pull/270

**Comment by splatter96** [comment]
Hello,
thanks for the feedback :)
Sorry for the late reply! Should I add the performance comparison to the source similarly as it is done in the official stable baselines3 algorithm pages? As in create a baselines3-zoo config for it and add the plots to this PR?

**Comment by araffin** [comment]
yes please =)

**Comment by daniel-redder** [comment]
> Hello, thanks for the feedback :) Sorry for the late reply! Should I add the performance comparison to the source similarly as it is done in the official stable baselines3 algorithm pages? As in create a baselines3-zoo config for it and add the plots to this PR?

Do you have the baseline results for this? I came across this PR looking for implementations on SACD. Thank you.

File: sb3_contrib/sacd/policies.py [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+        for idx in range(n_critics):
+            q_net_list = create_mlp(features_dim, num_actions.item(), net_arch, activation_fn)
+            q_net = nn.Sequential(*q_net_list)
+            self.add_module(f"qf{idx}", q_net)
+            self.q_networks.append(q_net)
+
+    def get_crit_params(self, n):
+        return self.q_networks[n].parameters()
+
+    def forward(self, obs: th.Tensor) -> Tuple[th.Tensor, ...]:
```
Comment (gyunt) [inline sb3_contrib/sacd/policies.py:194]: It seems that the case when self.features_extractor is None is not handled.
Would something like tuple(q_net(obs) for q_net in self.q_networks) work?

**Comment by araffin** [comment]
Hello,

sorry for the late reply.

> Plan:

That would be a good start. Please use the RL Zoo for that (and the wandb integration if possible).

> I already tested the PR locally on CartPole and it trains perfectly (reaches ~250 reward very fast).

250 is not very good for CartPole (it should reach ~500).
